fix(Popper): prevent flash of incorrectly positioned popper on open

fallmo opened the pull request:
## Description
Fixes an issue where popper elements (dropdowns, select, etc.) briefly flash at the top-left of the screen before appearing in their correct position.


https://github.com/user-attachments/assets/de09502e-94d7-4c39-9c46-6090e04321db


## Issue
When a popper opens, there's a race condition between:
1. The popper element being rendered (initially with `opacity: 0`)
2. Popper.js calculating and applying the correct position
3. The opacity transitioning to `1`

Previously, the opacity would transition to `1` immediately after setting `internalIsVisible`, which could occur before Popper.js finished calculating and applying the position transform.


## Fix
Added two `requestAnimationFrame` calls in the `show()` function:
- The first ensures React has committed DOM changes and the popper element exists
- The second ensures Popper.js has calculated positions and applied transforms

Only after both frames do we allow the opacity transition to begin. This guarantees the browser has painted the correctly positioned element before making it visible. 

## Testing
The issue occurs intermittently and can be hard to reproduce reliably.  
If you haven't encountered it, go to the Menus > Dropdown page, open and close the dropdowns back to back, then repeat with Menus > Select. The issue should reproduce within a few attempts.






<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved popper display timing so tooltips, dropdowns and popovers wait for layout and animation frames before becoming visible, yielding more accurate placement and smoother reveal/hide transitions.

* **Tests**
  * Updated unit and e2e tests to wait for popper visibility transitions and extended timeouts where needed, reducing flakiness and improving snapshot/visual stability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fallmo committed:
fix(Drawer): update refs in examples

Update ref types to HTMLSpanElement and initialize with null to fix TS errors

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

fallmo committed:
Merge branch 'patternfly:main' into main

fallmo committed:
fix(Popper): prevent flash of incorrectly positioned popper on open

Use requestAnimationFrame to ensure Popper.js has calculated and applied position transforms before transitioning opacity to 1.

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

patternfly-build commented:
Preview: https://pf-react-pr-12177.surge.sh

A11y report: https://pf-react-pr-12177-a11y.surge.sh

fallmo committed:
test: update snapshot tests to wait for popper opacity transition

Update Jest snapshot tests to wait for popper elements to complete
their opacity transition after the double requestAnimationFrame fix.
This ensures snapshots are taken after poppers are fully visible
(opacity: 1) rather than during the transition phase (opacity: 0).

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

fallmo committed:
fix(Popper): prevent race conditions by properly cleaning up animation frames

Add cleanup for requestAnimationFrame to prevent race conditions when popper
visibility changes rapidly or component unmounts during RAF execution.

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

fallmo committed:
fix(tests): increase tooltip visibility timeout to accomodate for async nature

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

nicolethoen reviewed (commented):
Overall I like this change but I do have two concerns/callouts:

- Any PF consumers who may have written unit tests related to testing popper dependent components are at risk of having their unit tests now fail. I _think_ that if they are in that situation, it'd indicate they are not following testing best practices, but that's the sort of thing I'd love to process with @wise-king-sullyman. It's also likely that if they were trying to test their popper components, that those test could be flakey themselves since they were needing to contend with the flashing issue as well... so this may be a net improvement for even the projects it affects. 🤔 

- I think searchinput.spec.ts also needs to be updated on line 56 to add the timeout

fallmo committed:
fix(tests): increase tooltip visibility timeout to accommodate async positioning

Signed-off-by: Mohamed Fall <ps.hackmaster@gmail.com>

wise-king-sullyman reviewed (changes requested):
Thanks for making this PR! I just have a few questions/concerns.

fallmo commented:
> Thanks for making this PR! I just have a few questions/concerns.

Hey, thanks for the review !

fallmo commented on packages/react-core/src/helpers/util.ts:
```diff
@@ -524,6 +524,17 @@ export const clearTimeouts = (timeoutRefs: React.RefObject<any>[]) => {
+    }
+  });
+};
+
```
Hey, you're absolutely right here. The `clearAnimationFrames` function takes an array, but in practice I only pass one value. I was following the pattern I saw with the `clearTimeouts` function, while also accounting for the possibility that there be varied uses of the function in the future.

fallmo commented on packages/react-integration/cypress/integration/button.spec.ts:
```diff
@@ -9,7 +9,8 @@ describe('Button Demo Test', () => {
     });
-    cy.get('.pf-v6-c-tooltip').should('be.visible');
+    // Tooltip visibility is async due to requestAnimationFrame-based positioning
+    cy.get('.pf-v6-c-tooltip', { timeout: 6000 }).should('be.visible');
```
Fair point. Given the current behavior this felt like the safest tradeoff, but I’d definitely welcome other suggestions.

wise-king-sullyman replied:
Are these test changes actually needed? I pulled your branch down and have run the tests a good few times with the timeouts removed and I'm not seeing any failures.

fallmo replied:
Right, in normal circumstances these changes aren't necessary; the raf calculations happen fast enough that a delay is pretty much non-existent. It's just that in many CI environments, the runners are super slow (cpu-constrained), as a result anything even slightly asynchronous may need a timeout.

fallmo replied:
I created an issue #12264 to document the problem a bit more.

wise-king-sullyman commented on packages/react-core/src/helpers/Popper/Popper.tsx:
```diff
@@ -234,6 +234,7 @@ export const Popper: React.FunctionComponent<PopperProps> = ({
   const transitionTimerRef = useRef(null);
   const showTimerRef = useRef(null);
   const hideTimerRef = useRef(null);
+  const rafRef = useRef<number>(null);
```
Hey sorry for the delay in circling back to this, yes I think that would be a better name.

fallmo committed:
fix(Popper): rename animation frame ref and adapt clearAnimationFrame function
